pkg/dns/target.go

```diff
@@ -149,34 +127,38 @@ func (t *ClusterGatewayTarget) GetWeight() int {
 }
 
 func (t *ClusterGatewayTarget) GetName() string {
-	return t.ClusterName
+	return t.Cluster.GetName()
 }
 
 func (t *ClusterGatewayTarget) GetShortCode() string {
 	return ToBase36hash(t.GetName())
 }
 
 func (t *ClusterGatewayTarget) setGeo(defaultGeo GeoCode) {
-	if defaultGeo == DefaultGeo || t.ClusterAttributes.Geo == nil {
-		t.Geo = &defaultGeo
+	geoCode := defaultGeo
+	if geoCode == DefaultGeo {
+		t.Geo = &geoCode
 		return
 	}
-	t.Geo = t.ClusterAttributes.Geo
+	if gc, ok := t.Cluster.GetLabels()[LabelLBAttributeGeoCode]; ok {
+		geoCode = GeoCode(gc)
+	}
+	t.Geo = &geoCode
 }
 
 func (t *ClusterGatewayTarget) setWeight(defaultWeight int, customWeights []*v1alpha1.CustomWeight) {
-	weight := &defaultWeight
-	if t.ClusterAttributes.CustomWeight != nil && customWeights != nil {
-		for k := range customWeights {
-			cw := customWeights[k]
-			if *t.ClusterAttributes.CustomWeight == cw.Value {
+	weight := defaultWeight
+	for k := range customWeights {
+		cw := customWeights[k]
+		if selector, err := metav1.LabelSelectorAsSelector(cw.Selector); err == nil {
```

Comment:
what does it mean if there is an error here? does it mean it is an invalid selector?

Reply:
I did dig into the code a bit to see if i could add a test that would actually cause this to throw an error, but it didn't really seem possible. It calls `labels.NewRequirement` internally and that will throw an error if the values are an incorrect length for the type of match you want to do, Exact in the case of Match Labels. These values will alwys be what is expected when they are passed though `LabelSelectorAsSelector` afaik.

Even if there is some way to cause this to throw an error it would be based on a misconfiguration of the selector and I'm not sure what else we can really do other than continue and use the default in that scenario.

Reply:
Found a way to make it fail. It checks the format of the labels and expects any with a / to have 2 parts, so this is an example of something that wouldn't work and would case an error here:

```
Selector: &metav1.LabelSelector{
   MatchLabels: map[string]string{
      "/tstlabel1": "TSTATTR",
   },
},
```

Added a test to with the above, but not really sure we can do anything other than ignore it. A log might be useful for debug purposes.

Reply:
hmm why would we not add something to the status to report that the label selector was wrong?

Reply:
Right. I guess we would need to bubble the error back up? The issue that I see is this is a public API and if I use it incorrectly, I would want to know what I have done wrong. With how it is now, it seems I would be left guessing? 
If we agree that the error should make its way to the status then this is something we could follow up with.

Reply:
I added the error https://github.com/Kuadrant/multicluster-gateway-controller/pull/462/commits/6e09f71650c14f872080a6898dfb8e43f4abc60d

Makes sense i guess since it's something they could mess up.

Reply:
``` status:
  conditions:
  - lastTransitionTime: "2023-08-31T09:34:39Z"
    message: 'reconcile DNSRecords error failed to create multi cluster gateway target
      for listener api : key: Invalid value: "/lb-attribute-custom-weight": prefix
      part must be non-empty '
    reason: ReconciliationError
    status: "False"
    type: Ready ```

Reply:
```bash
kubectl get gateway prod-web -n multi-cluster-gateways -o json | jq '.status.conditions[] | select(.type=="kuadrant.io/DNSPolicyAffected")'
{
  "lastTransitionTime": "2023-08-31T09:37:21Z",
  "message": "policy failed. Object unaffected by policy kuadrant.io/v1alpha1, Kind=DNSPolicy in namespace multi-cluster-gateways with name prod-web with error failed to create multi cluster gateway target for listener api : key: Invalid value: \"/lb-attribute-custom-weight\": prefix part must be non-empty ",
  "observedGeneration": 1,
  "reason": "Invalid",
  "status": "False",
  "type": "kuadrant.io/DNSPolicyAffected"
}

```